File: ai-rag-app/ai-rag-app/Networking/OshaAPIClient.swift
//
//  OshaAPIClient.swift
//  ai-rag-app
//
//  Created by Graham Diehl on 8/26/26.
//

import Foundation

enum OshaAPIError: Error {
    case invalidResponse
    case server(statusCode: Int, message: String)
    case decoding(Error)
    case transport(Error)

    var userMessage: String {
        switch self {
        case .invalidResponse:
            return "Received an unexpected response from the server."
        case .server(_, let message):
            return message
        case .decoding:
            return "Couldn't read the server's response."
        case .transport(let error):
            return error.localizedDescription
        }
    }
}

protocol OshaAPIClientProtocol {
    func askOshaPPE(query: String) async throws -> OshaStandardResponse
}

final class OshaAPIClient: OshaAPIClientProtocol {
    private let baseURL: URL
    private let session: URLSession

    init(baseURL: URL = URL(string: "http://localhost:8000")!, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
    }

    func askOshaPPE(query: String) async throws -> OshaStandardResponse {
        var request = URLRequest(url: baseURL.appendingPathComponent("/v1/osha/ppe"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(OshaStandardQuery(query: query, results: 5))

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw OshaAPIError.transport(error)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw OshaAPIError.invalidResponse
        }

        guard (200..<300).contains(httpResponse.statusCode) else {
            let message = (try? JSONDecoder().decode(OshaValidationError.self, from: data))?.detail
                ?? "Request failed with status \(httpResponse.statusCode)."
            throw OshaAPIError.server(statusCode: httpResponse.statusCode, message: message)
        }

        do {
            return try JSONDecoder().decode(OshaStandardResponse.self, from: data)
        } catch {
            throw OshaAPIError.decoding(error)
        }
    }
}

File: ai-rag-android/app/src/main/java/net/apperdashery/demo/ai_rag_android/network/OshaApiClient.kt
package net.apperdashery.demo.ai_rag_android.network

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import net.apperdashery.demo.ai_rag_android.model.OshaStandardQuery
import net.apperdashery.demo.ai_rag_android.model.OshaStandardResponse
import net.apperdashery.demo.ai_rag_android.model.OshaValidationError
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException

sealed class OshaApiException(message: String) : Exception(message) {
    class Server(val statusCode: Int, val userMessage: String) : OshaApiException(userMessage)
    class Transport(cause: Throwable) : OshaApiException(cause.message ?: "Network error")
    class Decoding(cause: Throwable) : OshaApiException("Couldn't read the server's response.")
}

interface OshaApiClient {
    suspend fun askOshaPPE(query: String): OshaStandardResponse
}

class OshaApiClientImpl(
    // 127.0.0.1 requires `adb reverse tcp:8000 tcp:8000` before launching the app.
    // (10.0.2.2, the emulator's usual host alias, was unreliable on this AVD image's
    // simulated Wi-Fi network — adb reverse tunnels over the adb transport instead
    // of the emulator's virtual NIC, so it isn't affected.)
    private val baseUrl: String = "http://127.0.0.1:8000",
    private val client: OkHttpClient = OkHttpClient(),
    private val json: Json = Json { ignoreUnknownKeys = true }
) : OshaApiClient {

    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    override suspend fun askOshaPPE(query: String): OshaStandardResponse = withContext(Dispatchers.IO) {
        val requestBody = json.encodeToString(
            OshaStandardQuery.serializer(),
            OshaStandardQuery(query = query)
        ).toRequestBody(jsonMediaType)

        val request = Request.Builder()
            .url("$baseUrl/v1/osha/ppe")
            .post(requestBody)
            .build()

        val response = try {
            client.newCall(request).execute()
        } catch (e: IOException) {
            throw OshaApiException.Transport(e)
        }

        response.use {
            val bodyString = it.body?.string().orEmpty()

            if (!it.isSuccessful) {
                val message = runCatching {
                    json.decodeFromString(OshaValidationError.serializer(), bodyString).detail
                }.getOrNull() ?: "Request failed with status ${it.code}."
                throw OshaApiException.Server(it.code, message)
            }

            try {
                json.decodeFromString(OshaStandardResponse.serializer(), bodyString)
            } catch (e: Exception) {
                throw OshaApiException.Decoding(e)
            }
        }
    }
}
